### PR TITLE
Improve moth breeding

### PR DESCRIPTION
**Pull request by OpenBagTwo**
Addresses #103, though it turns out that moths could *already be bred* using, um, [wheat](https://github.com/quiqueck/BetterEnd/issues/103#issuecomment-1296300947), so the primary thing this mod changes is explicitly setting the "breeding item" to any end plant.

### Validation Performed
- [x] Compiled the mod successfully
- [x] loaded an existing testing world, confirmed that older silk moths didn't glitch out
- [x] was unable to feed wheat to legacy silk moths or newly summoned ones
- [x] feeding `better_end.umbrella_moss` (but not `better_end.umbrella_moss_tall` to a silk moth caused it to enter "Love mode."
- [x] confirmed that this umbrella moss was the one that gets dropped from bonemealing naturally occurring umbrella moss
- [x] feeding two caused them to pursue the breeding goal and produce an offspring
  
- [x] immediately after, parent moths were unable to breed, feeding the "baby" caused it to emit "aging" particles rather than it entering love move (this implies that thee `Animal.isBaby` attribute is being correctly passed through)

### Note
- After breeding, the new silk moth had an Age value of (at least?) -20000 (need to re-validate, as I think the default is -24000)
- `better_end.umbrella_moss_tall` cannot be used to breed silk moths, only the 1-block-tall variant. The former is only obtainable by shearing a tall umbrella moss, the former is the drop from bonemealing. I'm not :100: clear on which items can be used for silk moth breeding (see in-line annotation for details). From empirical testing, it looks like saplings don't qualify, but Vaiolush Ferns do.
- I have made no changes to the AI. Explicitly: unlike bees, moths have no food-seeking goal.

### ~~Not~~ (also) Implemented
- [x] Baby silk moths should appear smaller than their parents

~~I'll add additional annotations in-line--I would love feedback on whether anything should be implemented differently.~~ (done)

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -29,6 +29,17 @@ repositories {
     maven { url "https://ladysnake.jfrog.io/artifactory/mods" }
     maven { url = "https://dvs1.progwml6.com/files/maven/" }
     maven { url = "https://modmaven.dev" }
+    exclusiveContent {
```
Comment (OpenBagTwo) [inline build.gradle:32]: Happy to revert this change, though it was really nice to not need to manually download BCLib.

File: src/main/java/org/betterx/betterend/entity/SilkMothEntity.java [hunk 2]
```diff
@@ -46,7 +51,7 @@
 import java.util.EnumSet;
 import org.jetbrains.annotations.Nullable;
 
-public class SilkMothEntity extends DespawnableAnimal implements FlyingAnimal {
```
Comment (OpenBagTwo) [inline src/main/java/org/betterx/betterend/entity/SilkMothEntity.java:49]: Presumably if you're making a silk moth farm, having to nametag all your moths to prevent them from despawning would be a pain in the toucans.

File: src/main/java/org/betterx/betterend/entity/SilkMothEntity.java [hunk 3]
```diff
@@ -76,11 +81,12 @@ public void setHive(Level world, BlockPos hive) {
 
     @Override
     public boolean canBeLeashed(Player player) {
-        return false;
+        return true;
```
Comment (OpenBagTwo) [inline src/main/java/org/betterx/betterend/entity/SilkMothEntity.java:84]: I don't feel strongly about this modification, though I did validate that it's working:



 I didn't validate whether they'll follow you if you're holding food, and I'm actually not at all sure if a silk touched hive can be used to transport silk moths, so having this as a method to move your moths around seems like it would be a nice QoL improvement.

File: src/main/java/org/betterx/betterend/entity/SilkMothEntity.java [hunk 4]
```diff
@@ -76,11 +81,12 @@ public void setHive(Level world, BlockPos hive) {
...
     }
 
     @Override
     public void addAdditionalSaveData(CompoundTag tag) {
+        super.addAdditionalSaveData(tag);
```
Comment (OpenBagTwo) [inline src/main/java/org/betterx/betterend/entity/SilkMothEntity.java:89]: I'm not sure if it was intentional not to call the super method, but *I believe* that without this, not even custom names would be saved (see: despawnable).

File: src/main/java/org/betterx/betterend/entity/SilkMothEntity.java [hunk 5]
```diff
@@ -155,7 +162,7 @@ public boolean isNoGravity() {
 
     @Override
     public AgeableMob getBreedOffspring(ServerLevel world, AgeableMob entity) {
-        return EndEntities.SILK_MOTH.create(world);
+        return SILK_MOTH.create(world);
```
Comment (OpenBagTwo) [inline src/main/java/org/betterx/betterend/entity/SilkMothEntity.java]: Accidental modification--happy to revert.

File: src/main/java/org/betterx/betterend/entity/SilkMothEntity.java [hunk 6]
```diff
@@ -322,4 +329,17 @@ public void tick() {
             }
         }
     }
+
+    @Override
+    public boolean isFood(ItemStack itemStack) {
+        if (!(itemStack.getItem() instanceof BlockItem)){
+            return false;
+        }
+        return ((BlockItem) (itemStack.getItem())).getBlock() instanceof EndPlantBlock;
```
Comment (OpenBagTwo) [inline src/main/java/org/betterx/betterend/entity/SilkMothEntity.java:338]: #103 suggested flowers be used to breed moths. Given that moths don't actually seek out end flowers--and that there's no dedicated list of end "flowers", I just used any end vegetation. Seemed like a least-effort solution, but I could see either:

- Creating a hard-coded list of "end flowers"
- Dynamically determining whether an end plant is a flower based on whether it can be crafted into dye
- Giving silk moths a dedicated breeding item, such as the [blossom berry](https://github.com/quiqueck/BetterEnd/issues/103#issuecomment-1296936348)
- Letting silk moths be bred by any _overworld_ flower instead (cool lore implications?) / as well
- A nod towards [their biological counterparts](https://www.dkfindout.com/uk/animals-and-nature/insects/silk-moth/) would be to have them eat leaves, either specific to a single tree or generically any.

File: src/main/java/org/betterx/betterend/entity/SilkMothEntity.java [hunk 7]
```diff
@@ -322,4 +329,17 @@ public void tick() {
...
+    }
+
+    @Override
+    public InteractionResult mobInteract(Player player, InteractionHand interactionHand) {
+        return super.mobInteract(player, interactionHand);
```
Comment (OpenBagTwo) [inline src/main/java/org/betterx/betterend/entity/SilkMothEntity.java:343]: Obviously this doesn't do anything, so happy to revert.

**Comment by OpenBagTwo** [comment]
Got the babies rendering.


As you can see, I didn't bother with the hitboxes.

**Comment by ChristopherJ123** [comment]
Seems to be working just fine!



I'm not sure but I think Silk moth nest are a bit broken, they don't behave the same with bee nest. I think they don't store how many silk moth's are in the nest when broken with silk touch, because they would stack in the inventory. Destroying a silk moth nest without silk touch also doesn't remove the silk moths inside. P.S Sorry bad english!

**Comment by OpenBagTwo** [comment]
> I'm not sure but I think Silk moth nest are a bit broken, they don't behave the same with bee nest.

Let's spin this off into a separate issue (bug report). If you can, test both Silk Moth Hives (!) and Silk Moth Nests, as it looks like the two blocks utilize completely separate classes with (at least slightly) different code.

**Update:** just a quick note that in the current Minecraft version `Blocks.BEEHIVE` and `Blocks.BEE_NEST` are both instances of the `BeehiveBlock` class.

File: build.gradle (hunk 1 above)
Comment (quiqueck) [reply build.gradle:32]: Does this pull the correct files? I do not upload dev-versions of BCLib to modrinth...

**Comment by quiqueck** [comment]
Wow, thanks for all the work. :D
